client/src/components/habits/calendar/synthetic.ts

```diff
@@ -0,0 +1,90 @@
+import { isNullish } from "@shared/lib/is-nullish";
+import type {
+	HabitWithPossiblySyntheticEntries,
+	PossiblySyntheticHabitEntry,
+} from "@shared/lib/schemas/habit";
+import type { Datelike } from "@shared/lib/schemas/timestamp";
+import { isSynthetic } from "@shared/types/data/habit-entry.guards";
+import type { Nullable } from "@shared/types/data/utility.types";
+import type { Dayjs } from "dayjs";
+import { createDate } from "@/lib/datetime/make-date";
+import { makeSyntheticEntry } from "../Habits/synthetic";
+
+/** Compute the number of _real_ entries required for successful habit
+ * completion on a given day, for the specific use-case of the habit calendar
+ * (which differs in business logic from the completions in Today etc.) */
+export function realEntryThreshold(
+	habit: HabitWithPossiblySyntheticEntries,
+	date: Dayjs
+) {
+	if (
+		date.isBefore(createDate(habit.start_timestamp), "date") ||
+		(habit.end_timestamp &&
+			date.isAfter(createDate(habit.end_timestamp), "date"))
+	) {
+		return 0;
+	}
+
+	switch (habit.interval_unit) {
+		case "day": {
+			return habit.frequency;
+		}
+		// Weekly, monthly, [yearly: not yet implemented] should assume a
+		// single entry per day. Schemes like 20x/week are exceptions that we
```

Comment:
I think I have an issue that's related to yearly intervals and such. Probably best to include the issue ID in this TODO.